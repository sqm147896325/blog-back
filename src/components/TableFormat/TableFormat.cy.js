/// <reference types="Cypress" />
import TableFormat from './TableFormat.vue'

describe('<TableFormat />', () => {
  it('renders', () => {
    // see: https://on.cypress.io/mounting-vue
    cy.viewport(1024, 768)

    cy.mount(TableFormat, {
      propsData: {
        modelValue: true,
        tableOption: [
          { field: 'title', cname: '标题', cshow: true, align: 'center', showname: '标题', clock: false, width: 180 },
          { field: 'author', cname: '作者名', cshow: true, align: 'center', showname: '作者名', clock: false, width: 100 },
          { field: 'author_id', cname: '作者id', cshow: true, align: 'center', showname: '作者id', clock: false, width: 80 },
          { field: 'des', cname: '描述', cshow: true, align: 'center', showname: '描述', clock: false, width: 160 },
          { field: 'keyword', cname: '关键字', cshow: true, align: 'center', showname: '关键字', clock: false, width: 160 },
          { field: 'lenght', cname: '字数', cshow: true, align: 'center', showname: '字数', clock: false, width: 160 },
          { field: 'visited', cname: '访问次数', cshow: true, align: 'center', showname: '访问次数', clock: false, width: 160 },
          { field: 'created_at', cname: '创建时间', cshow: true, align: 'center', showname: '创建时间', clock: false, width: 'auto' },
          { field: 'updated_at', cname: '更新时间', cshow: true, align: 'center', showname: '更新时间', clock: false, width: 'auto' },
          { field: 'operation', cname: '操作', cshow: true, align: 'center', showname: '操作', clock: true, width: 270 }
        ]
      }
    })
  })

  it('重置方法', () => {
    cy.mount(TableFormat, {
      propsData: {
        modelValue: true,
        tableOption: [
          { field: 'title', cname: '标题', cshow: true, align: 'center', showname: '标题', clock: false, width: 180 },
          { field: 'author', cname: '作者名', cshow: true, align: 'center', showname: '作者名', clock: false, width: 100 }
        ]
      }
    })

    cy.get('tbody > :nth-child(1) > .el-table_2_column_8 > .cell')
      .contains('标题')
      .should('exist')

    cy.get(':nth-child(1) > .el-table_2_column_12 > .cell > .el-checkbox')
      .click()

    cy.contains('下移')
      .click()

    cy.get('tbody > :nth-child(1) > .el-table_2_column_8 > .cell')
      .contains('作者名')
      .should('exist')

    cy.contains('重置')
      .click()

    cy.get('tbody > :nth-child(1) > .el-table_2_column_8 > .cell')
      .contains('标题')
      .should('exist')
  })
})
