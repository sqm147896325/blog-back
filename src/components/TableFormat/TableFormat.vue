<template>
  <el-dialog
    title="格式"
    v-bind="$attrs"
    width="560px"
  >
    <div class="format-row">
      <el-button
        type="primary"
        @click="sure"
      >
        确定
      </el-button>
      <el-button
        type="primary"
        @click="reset"
      >
        重置
      </el-button>
      <el-button
        type="primary"
        :disabled="currIndex === null"
        @click="move('top')"
      >
        最上
      </el-button>
      <el-button
        type="primary"
        :disabled="currIndex === null"
        @click="move('up')"
      >
        上移
      </el-button>
      <el-button
        type="primary"
        :disabled="currIndex === null"
        @click="move('down')"
      >
        下移
      </el-button>
      <el-button
        type="primary"
        :disabled="currIndex === null"
        @click="move('lowest')"
      >
        最下
      </el-button>
    </div>
    <div class="format-list">
      <el-table
        :data="tableData"
        stripe
        border
        height="300"
        highlight-current-row
        :row-class-name="tableRowIndex"
        @current-change="handleCurrentChange"
      >
        <el-table-column
          align="center"
          header-align="center"
          type="index"
          width="50"
          label="序号"
        />
        <el-table-column
          align="center"
          header-align="center"
          prop="cname"
          label="名称"
          width="100"
        />
        <el-table-column
          align="center"
          header-align="center"
          prop="cshow"
          label="显示"
          width="50"
        >
          <template #default="{ row }">
            <el-checkbox v-model="row.cshow" />
          </template>
        </el-table-column>
        <el-table-column
          align="center"
          header-align="center"
          prop="align"
          label="对齐方式"
          width="auto"
        >
          <template #default="{ row }">
            <el-select
              v-model="row.align"
            >
              <el-option
                v-for="item in alignList"
                :key="item.value"
                :label="item.label"
                :value="item.value"
              />
            </el-select>
          </template>
        </el-table-column>
        <el-table-column
          align="center"
          header-align="center"
          prop="showname"
          label="显示名称"
          width="120"
        >
          <template #default="{ row }">
            <el-input
              v-model="row.showname"
            />
          </template>
        </el-table-column>
        <el-table-column
          align="center"
          header-align="center"
          prop="clock"
          label="锁定"
          width="50"
        >
          <template #default="{ row }">
            <el-checkbox v-model="row.clock" />
          </template>
        </el-table-column>
      </el-table>
    </div>
  </el-dialog>
</template>

<script>
export default {
  name: 'TableFormat',
  props: {
    tableOption: {
      type: Array,
      default: () => []
    }
  },
  emits: ['setTableOption', 'update:modelValue'],
  data () {
    return {
      alignList: [
        { value: 'left', label: '左对齐' },
        { value: 'center', label: '居中' },
        { value: 'right', label: '右对齐' }
      ],
      tableData: [],
      rawData: [], // 存放原始数据
      currIndex: null // 控制上移下移的高亮行索引
    }
  },
  watch: {
    // 表格数据是对tableOption进行处理
    tableOption (newValue) {
      this.dataChange(newValue)
    }
  },
  mounted () {
    this.dataChange(this.tableOption)
  },
  methods: {
    sure () {
      const tableData = [...this.tableData].map(e => {
        e = JSON.parse(JSON.stringify(e))
        delete e.index
        return e
      })
      this.$emit('setTableOption', tableData)
      this.$emit('update:modelValue', false)
    },
    reset () {
      this.currIndex = null
      this.tableData = this.rawData.map(e => {
        const tableRow = JSON.parse(JSON.stringify(e))
        return tableRow
      })
    },
    move (type) {
      const index = this.currIndex
      switch (type) {
        case 'top':
          if (index !== 0) {
            this.tableData.unshift(this.tableData.splice(index, 1)[0])
            this.currIndex = 0
          }
          break
        case 'up':
          if (index !== 0) {
            this.tableData[index] = this.tableData.splice(index - 1, 1, this.tableData[index])[0]
            this.currIndex = this.currIndex - 1
          } else {
            this.tableData.push(this.tableData.shift())
            this.currIndex = this.tableData.length - 1
          }
          break
        case 'down':
          if (index !== this.tableData.length - 1) {
            this.tableData[index] = this.tableData.splice(index + 1, 1, this.tableData[index])[0]
            this.currIndex = this.currIndex + 1
          } else {
            this.tableData.unshift(this.tableData.splice(index, 1)[0])
            this.currIndex = 0
          }
          break
        case 'lowest':
          if (index !== this.tableData.length - 1) {
            this.tableData.push(this.tableData.splice(index, 1)[0])
            this.currIndex = this.tableData.length - 1
          }
          break
        default:
          break
      }
    },
    /* 表格方法 */
    // 高亮行改变
    handleCurrentChange (currentRow) {
      if (currentRow && currentRow.index !== null) {
        this.currIndex = currentRow.index
      }
    },
    // 当前行index
    tableRowIndex ({ row, rowIndex }) {
      row.index = rowIndex
    },
    /* 数据变化监听 */
    dataChange (newValue) {
      this.rawData = newValue.map((e, i) => {
        const tableRow = JSON.parse(JSON.stringify(e))
        return tableRow
      })
      this.tableData = newValue.map((e, i) => {
        const tableRow = JSON.parse(JSON.stringify(e))
        return tableRow
      })
    }
  }
}
</script>

<style lang="less" scoped>
.format-row{
    display: flex;
    margin-bottom: 10px;
}
.format-list{
    height: 320px;
    overflow-y: scroll;
}
</style>
