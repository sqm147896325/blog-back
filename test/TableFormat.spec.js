import { mount } from '@vue/test-utils'

import ElementPlus from 'element-plus'
import * as ElementPlusIconsVue from '@element-plus/icons-vue'
import 'element-plus/dist/index.css' // 导入全局样式

import TableFormat from '../src/components/TableFormat/TableFormat.vue'

describe('TableFormat', () => {
  it('组件挂载', () => {
    const wrapper = mount(TableFormat, {
      global: {
        plugins: [ElementPlus, ElementPlusIconsVue]
      },
      props: {
        modelValue: true,
        tableOption: [
          { field: 'title', cname: '标题', cshow: true, align: 'center', showname: '标题', clock: false, width: 180 },
          { field: 'author', cname: '作者名', cshow: true, align: 'center', showname: '作者名', clock: false, width: 100 },
          { field: 'author_id', cname: '作者id', cshow: true, align: 'center', showname: '作者id', clock: false, width: 80 },
          { field: 'des', cname: '描述', cshow: true, align: 'center', showname: '描述', clock: false, width: 160 },
          { field: 'keyword', cname: '关键字', cshow: true, align: 'center', showname: '关键字', clock: false, width: 160 },
          { field: 'lenght', cname: '字数', cshow: true, align: 'center', showname: '字数', clock: false, width: 160 },
          { field: 'visited', cname: '访问次数', cshow: true, align: 'center', showname: '访问次数', clock: false, width: 160 },
          { field: 'created_at', cname: '创建时间', cshow: true, align: 'center', showname: '创建时间', clock: false, width: 'auto' },
          { field: 'updated_at', cname: '更新时间', cshow: true, align: 'center', showname: '更新时间', clock: false, width: 'auto' },
          { field: 'operation', cname: '操作', cshow: true, align: 'center', showname: '操作', clock: true, width: 270 }
        ]
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('确定方法', () => {
    const wrapper = mount(TableFormat, {
      global: {
        plugins: [ElementPlus, ElementPlusIconsVue]
      },
      props: {
        modelValue: true,
        tableOption: [
          { field: 'title', cname: '标题', cshow: true, align: 'center', showname: '标题', clock: false, width: 180 },
          { field: 'author', cname: '作者名', cshow: true, align: 'center', showname: '作者名', clock: false, width: 100 },
          { field: 'author_id', cname: '作者id', cshow: true, align: 'center', showname: '作者id', clock: false, width: 80 },
          { field: 'des', cname: '描述', cshow: true, align: 'center', showname: '描述', clock: false, width: 160 },
          { field: 'keyword', cname: '关键字', cshow: true, align: 'center', showname: '关键字', clock: false, width: 160 },
          { field: 'lenght', cname: '字数', cshow: true, align: 'center', showname: '字数', clock: false, width: 160 },
          { field: 'visited', cname: '访问次数', cshow: true, align: 'center', showname: '访问次数', clock: false, width: 160 },
          { field: 'created_at', cname: '创建时间', cshow: true, align: 'center', showname: '创建时间', clock: false, width: 'auto' },
          { field: 'updated_at', cname: '更新时间', cshow: true, align: 'center', showname: '更新时间', clock: false, width: 'auto' },
          { field: 'operation', cname: '操作', cshow: true, align: 'center', showname: '操作', clock: true, width: 270 }
        ]
      }
    })

    wrapper.vm.$nextTick(() => {
      // 模拟点击确定按钮
      wrapper.find('.format-row button:nth-child(1)').trigger('click')

      // 检查是否触发了setTableOption事件
      expect(wrapper.emitted('setTableOption')).toBeTruthy()

      // 检查是否触发了update:modelValue事件
      expect(wrapper.emitted('update:modelValue')).toBeTruthy()
    })
  })

  it('重置方法', () => {
    const wrapper = mount(TableFormat, {
      global: {
        plugins: [ElementPlus, ElementPlusIconsVue]
      },
      props: {
        modelValue: true,
        tableOption: [
          { field: 'title', cname: '标题', cshow: true, align: 'center', showname: '标题', clock: false, width: 180 },
          { field: 'author', cname: '作者名', cshow: true, align: 'center', showname: '作者名', clock: false, width: 100 },
          { field: 'author_id', cname: '作者id', cshow: true, align: 'center', showname: '作者id', clock: false, width: 80 },
          { field: 'des', cname: '描述', cshow: true, align: 'center', showname: '描述', clock: false, width: 160 },
          { field: 'keyword', cname: '关键字', cshow: true, align: 'center', showname: '关键字', clock: false, width: 160 },
          { field: 'lenght', cname: '字数', cshow: true, align: 'center', showname: '字数', clock: false, width: 160 },
          { field: 'visited', cname: '访问次数', cshow: true, align: 'center', showname: '访问次数', clock: false, width: 160 },
          { field: 'created_at', cname: '创建时间', cshow: true, align: 'center', showname: '创建时间', clock: false, width: 'auto' },
          { field: 'updated_at', cname: '更新时间', cshow: true, align: 'center', showname: '更新时间', clock: false, width: 'auto' },
          { field: 'operation', cname: '操作', cshow: true, align: 'center', showname: '操作', clock: true, width: 270 }
        ]
      }
    })

    wrapper.vm.$nextTick(() => {
      // 修改数据
      wrapper.vm.tableData = [{ cname: 'Name 1', cshow: false }]

      // 模拟点击重置按钮
      wrapper.find('.format-row button:nth-child(2)').trigger('click')

      // 检查数据是否重置成功
      expect(wrapper.vm.tableData).toEqual([
        { field: 'title', cname: '标题', cshow: true, align: 'center', showname: '标题', clock: false, width: 180 },
        { field: 'author', cname: '作者名', cshow: true, align: 'center', showname: '作者名', clock: false, width: 100 },
        { field: 'author_id', cname: '作者id', cshow: true, align: 'center', showname: '作者id', clock: false, width: 80 },
        { field: 'des', cname: '描述', cshow: true, align: 'center', showname: '描述', clock: false, width: 160 },
        { field: 'keyword', cname: '关键字', cshow: true, align: 'center', showname: '关键字', clock: false, width: 160 },
        { field: 'lenght', cname: '字数', cshow: true, align: 'center', showname: '字数', clock: false, width: 160 },
        { field: 'visited', cname: '访问次数', cshow: true, align: 'center', showname: '访问次数', clock: false, width: 160 },
        { field: 'created_at', cname: '创建时间', cshow: true, align: 'center', showname: '创建时间', clock: false, width: 'auto' },
        { field: 'updated_at', cname: '更新时间', cshow: true, align: 'center', showname: '更新时间', clock: false, width: 'auto' },
        { field: 'operation', cname: '操作', cshow: true, align: 'center', showname: '操作', clock: true, width: 270 }
      ])
    })
  })

  // 其他测试用例...
})
