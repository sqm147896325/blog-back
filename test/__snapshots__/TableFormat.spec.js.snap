// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableFormat > renders correctly 1`] = `<!--teleport start-->`;

exports[`TableFormat > 组件挂载 1`] = `<!--teleport start-->`;
